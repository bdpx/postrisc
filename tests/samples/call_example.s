.text

call_code_target:

.rodata
call_data_target:

.text
    jmp	callexample
;*****************************************************************
; Function  compute A**4 of parameter A, passed in register r33
;*****************************************************************
quadrat:
    write	"function quadrat entered: r0=%x128(r0)"
    alloc	93
    write	"rsc     %s(rsc)"
    write	"psr     %s(psr)"
    write	"rsc     %s(rsc)"
    mul	%r33, %r33, %r33
    mul	%r33, %r33, %r33
    write	"r0=%x128(r0) r33=%i64(r33)"
    write	"%m(dump)"
;	mtspr	%r45, psr
    write	"function quadrat exited"
    ret
end_quadrat:

;*****************************************************************
; Example of calling sequence with branch prediction
callexample:
    alloc	91
    ldi.l	%r90, 0x1234567890abcdef
    write	"arg3 %x64(r90)"
    srpi	%r89, %r90, %r90, 16
    write	"arg2 %x64(r89)"
    srpi	%r88, %r90, %r90, 16
    write	"arg1 %x64(r88)"
    ldi		%r87, 7		; setup arguments
;   write	"%m(dump)"
    write	"rsc: %s(rsc)"
    write	"function quadrat called"
    call.r	%r86, quadrat
    write	"rsc: %s(rsc)"
; Rest instructions after return from subroutine
;*****************************************************************
.text	; return to code section

; Here we test registers used by ABI (application binary interface)
; Check loader.
    write	"sp=%x64(sp) tp=%x64(tp) r0=%x128(r0)"
    write	"rsc: %s(rsc)"
    write	"psr: %s(psr)"
    write	"r14: %x64(r14)"
    write	"reta: %i64(r72)"		; out return address
    write	"retv: %i64(r73)"		; out return value
    write	"rsc: %s(rsc)"
    write	"rsc: %s(psr)"
    ldi.l	%r11, 0x407d8bffffccccff
    write	"r11: %x64(r11)"
    addi.l	%r12, %r11, 0x400000
    write	"r12: %x64(r12)"
    xor		%r20, %r19, %r11
    addi.l	%r20, %r20, 0x400000
    ldi		%r10, 10
    ldi		%r11, 11
    cmps.lt.d	%r2, %r11, %r10
    write	"%i64(r11) %i64(r10)"
    jmp		call_exit

    call.r	%r42, quadrat
    call.ri	%r42, %r34, %gz
    call.mi	%r42, %r34, 468
    call.plt	%r42, call_data_target
    call.ri	%r42, %r34, %gz

call_exit:
    write	"end call test"

.end
