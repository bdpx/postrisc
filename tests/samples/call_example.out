arg3 1234567890abcdef
arg2 cdef1234567890ab
arg1 cdef1234567890ab
rsc: 000000000000005b sof=91 bof=0 soc=0 ndirty=0 bod=0
function quadrat called
function quadrat entered: r0=000000000002c2d6_fffffffff1230090
rsc     000000002200565d sof=93 bof=86 soc=0 ndirty=34 bod=52
psr     0000000000010000 pm=0 id=0 dd=0 lp=0 tb=0 ss=0 ic=0 ib=0 pl=0 pp=0 ri=1 future=00000000000000000000000000000000
rsc     000000002200565d sof=93 bof=86 soc=0 ndirty=34 bod=52
r0=000000000002c2d6_fffffffff1230090 r33=0
special registers:
ip      fffffffff1230030
eip     fffffffff1230140
fpcr    0000000000000000 fe=vzoux rm=0
eca     0000000000000000
rsc     000000002200565d sof=93 bof=86 soc=0 ndirty=34 bod=52
rsp     fffffffff1730340
bsp     fffffffff1630000
psr     0000000000030000 pm=0 id=0 dd=0 lp=0 tb=0 ss=0 ic=0 ib=0 pl=0 pp=0 ri=3 future=00000000000000000000000000000000
reip    fffffffff1230140
kip     ffffffff00000000
ksp     0000000000000000
krsp    0000000000000000
peb     0000000000000000
teb     0000000000000000
itc     0000000000000008
itm     0000000000000000
pta     00000012345001c7 v=1 pta=0000001234500000 pageoffset=14 nlevels=3
iva     0000000000000000
iip     fffffffff1230000
iipa    fffffffff1230030
ipsr    0000000000000000 pm=0 id=0 dd=0 lp=0 tb=0 ss=0 ic=0 ib=0 pl=0 pp=0 ri=0 future=00000000000000000000000000000000
cause   0000000000000000 x=0 w=0 r=0 a=0 n=0 d=0 ei=0 code=0 vector=0
ifa     0000000000000000
iib     0000000000000000_0000000000000000
irr0    0000000000000000
irr1    0000000000000000
irr2    0000000000000000
irr3    0000000000000000
isr0    0000000000000000
isr1    0000000000000000
isr2    0000000000000000
isr3    0000000000000000
iv      0000000000000000
lid     0000000000000000
tpr     00000000000000ff vector=255
itcv    0000000000008000 vector=0 mask=1
tsv     0000000000008000 vector=0 mask=1
pmv     0000000000008000 vector=0 mask=1
cmcv    0000000000008000 vector=0 mask=1

general registers:
   r0 000000000002c2d6_fffffffff1230090   r1 0000000000000000_0000000000000007   r2 0000000000000000_cdef1234567890ab   r3 0000000000000000_cdef1234567890ab
   r4 0000000000000000_1234567890abcdef   r5 0000000000000000_0000000000000000   r6 0000000000000000_0000000000000000   r7 0000000000000000_0000000000000000
   r8 0000000000000000_0000000000000000   r9 0000000000000000_0000000000000000  r10 0000000000000000_0000000000000000  r11 0000000000000000_0000000000000000
  r12 0000000000000000_0000000000000000  r13 0000000000000000_0000000000000000  r14 0000000000000000_0000000000000000  r15 0000000000000000_0000000000000000
  r16 0000000000000000_0000000000000000  r17 0000000000000000_0000000000000000  r18 0000000000000000_0000000000000000  r19 0000000000000000_0000000000000000
  r20 0000000000000000_0000000000000000  r21 0000000000000000_0000000000000000  r22 0000000000000000_0000000000000000  r23 0000000000000000_0000000000000000
  r24 0000000000000000_0000000000000000  r25 0000000000000000_0000000000000000  r26 0000000000000000_0000000000000000  r27 0000000000000000_0000000000000000
  r28 0000000000000000_0000000000000000  r29 0000000000000000_0000000000000000  r30 0000000000000000_0000000000000000  r31 0000000000000000_0000000000000000
  r32 0000000000000000_0000000000000000  r33 0000000000000000_0000000000000000  r34 0000000000000000_0000000000000000  r35 0000000000000000_0000000000000000
  r36 0000000000000000_0000000000000000  r37 0000000000000000_0000000000000000  r38 0000000000000000_0000000000000000  r39 0000000000000000_0000000000000000
  r40 0000000000000000_0000000000000000  r41 0000000000000000_0000000000000000  r42 0000000000000000_0000000000000000  r43 0000000000000000_0000000000000000
  r44 0000000000000000_0000000000000000  r45 0000000000000000_0000000000000000  r46 0000000000000000_0000000000000000  r47 0000000000000000_0000000000000000
  r48 0000000000000000_0000000000000000  r49 0000000000000000_0000000000000000  r50 0000000000000000_0000000000000000  r51 0000000000000000_0000000000000000
  r52 0000000000000000_0000000000000000  r53 0000000000000000_0000000000000000  r54 0000000000000000_0000000000000000  r55 0000000000000000_0000000000000000
  r56 0000000000000000_0000000000000000  r57 0000000000000000_0000000000000000  r58 0000000000000000_0000000000000000  r59 0000000000000000_0000000000000000
  r60 0000000000000000_0000000000000000  r61 0000000000000000_0000000000000000  r62 0000000000000000_0000000000000000  r63 0000000000000000_0000000000000000
  r64 0000000000000000_0000000000000000  r65 0000000000000000_0000000000000000  r66 0000000000000000_0000000000000000  r67 0000000000000000_0000000000000000
  r68 0000000000000000_0000000000000000  r69 0000000000000000_0000000000000000  r70 0000000000000000_0000000000000000  r71 0000000000000000_0000000000000000
  r72 0000000000000000_0000000000000000  r73 0000000000000000_0000000000000000  r74 0000000000000000_0000000000000000  r75 0000000000000000_0000000000000000
  r76 0000000000000000_0000000000000000  r77 0000000000000000_0000000000000000  r78 0000000000000000_0000000000000000  r79 0000000000000000_0000000000000000
  r80 0000000000000000_0000000000000000  r81 0000000000000000_0000000000000000  r82 0000000000000000_0000000000000000  r83 0000000000000000_0000000000000000
  r84 0000000000000000_0000000000000000  r85 0000000000000000_0000000000000000  r86 0000000000000000_0000000000000000  r87 0000000000000000_0000000000000000
  r88 0000000000000000_0000000000000000  r89 0000000000000000_0000000000000000  r90 0000000000000000_0000000000000000  r91 0000000000000000_0000000000000000
  r92 0000000000000000_0000000000000000  r93 0000000000000000_0000000000000000  r94 0000000000000000_0000000000000000  r95 0000000000000000_0000000000000000
  r96 0000000000000000_0000000000000000  r97 0000000000000000_0000000000000000  r98 0000000000000000_0000000000000000  r99 0000000000000000_0000000000000000
 r100 0000000000000000_0000000000000000 r101 0000000000000000_0000000000000000 r102 0000000000000000_0000000000000000 r103 0000000000000000_0000000000000000
 r104 0000000000000000_0000000000000000 r105 0000000000000000_0000000000000000 r106 0000000000000000_0000000000000000 r107 0000000000000000_0000000000000000
 r108 0000000000000000_0000000000000000 r109 0000000000000000_0000000000000000 r110 0000000000000000_0000000000000000 r111 0000000000000000_0000000000000000
 r112 0000000000000000_0000000000000000 r113 0000000000000000_0000000000000000 r114 0000000000000000_0000000000000000 r115 0000000000000000_0000000000000000
 r116 0000000000000000_0000000000000000 r117 0000000000000000_0000000000000000 r118 0000000000000000_0000000000000000 r119 0000000000000000_0000000000000000
  --- 0000000000000000_0000000000000000  --- 0000000000000000_0000000000000000  --- 0000000000000000_0000000000000000  --- 0000000000000000_0000000000000000
  --- 0000000000000000_0000000000000000  --- 0000000000000000_0000000000000000  --- 0000000000000000_0000000000000000  --- 0000000000000000_0000000000000000
   g0 0000000000000000_0000000000000000   g1 0000000000000000_0000000000000000   g2 0000000000000000_0000000000000000   g3 0000000000000000_0000000000000000
   tp 0000000000000000_fffffffff1830000   fp 0000000000000000_fffffffff1640000   sp 0000000000000000_fffffffff1640000   gz 0000000000000000_0000000000000000

cpuid:
  size                   => 11
  version                => 0000000000000001
  capabilities           => 0000000000000000
  L1 icache              => 0000000080000440 rowsize=64 assoc=4 size=32768
  L1 dcache              => 0000000080000440 rowsize=64 assoc=4 size=32768
  L2 dcache              => 0000000400000840 rowsize=64 assoc=8 size=262144
  L3 dcache              => 0000002000000840 rowsize=64 assoc=8 size=2097152
  L1 itlb                => 0000000000000040 size=64
  L1 dtlb                => 0000000000000040 size=64
  L2 tlb                 => 0000000000000100 size=256
  PMR_MAX                => 20


function quadrat exited
rsc: 000000000000005b sof=91 bof=0 soc=0 ndirty=0 bod=0
sp=fffffffff1640000 tp=fffffffff1830000 r0=0000000000000000_0000000000000000
rsc: 000000000000005b sof=91 bof=0 soc=0 ndirty=0 bod=0
psr: 0000000000010000 pm=0 id=0 dd=0 lp=0 tb=0 ss=0 ic=0 ib=0 pl=0 pp=0 ri=1 future=00000000000000000000000000000000
r14: 0000000000000000
reta: 0
retv: 0
rsc: 000000000000005b sof=91 bof=0 soc=0 ndirty=0 bod=0
rsc: 0000000000030000 pm=0 id=0 dd=0 lp=0 tb=0 ss=0 ic=0 ib=0 pl=0 pp=0 ri=3 future=00000000000000000000000000000000
r11: 407d8bffffccccff
r12: 407d8c00000cccff
11 10
end call test
